sympy/series/series.py

```diff
@@ -61,3 +62,97 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    mathematician and teacher Hans Heinrich Bürmann ( --1817), both in the late 18th
+    century.
+
+    Theorem: Suppose x is defined as a function of y by an equation of the form
+
+    f(x) = y
+
+    where f is analytic at a point a and f'(a) is not equal to 0.
+    Then it is possible to invert or solve the equation for x in the form of a series:
+
+    .. math::
+        g(y) = a + \sum_{n = 1}^{oo} \frac{g_n (y - f(a))^n}{n!}
+
+    where
+
+    .. math::
+        g_n = \lim_{y \to a} \left[ \frac{d^{n-1}}{dx^{n-1}}
+        \left( \frac{x-a}{f(x)-f(a)} \right) ^n \right]
+
+    The theorem further states that this series has a non-zero radius of convergence,
+    i.e., x represents an analytic function of y in a neighbourhood of y = f(a).
+
+    Parameters
+    ==========
+
+    f : expression, the expansion of whose inverse is to be found.
+
+    x : variable, in terms of which equation is given, and in terms of which, the
+    expansion will be provided.
+
+    a : point on x-axis in the neighbourhood of which, the inverse expansion is to be found.
+
+    n : no of terms required in the expansion; default value is 3 terms.
+
+    Examples
+    ========
+
+    >>> from sympy.series.series import inversion_series
+    >>> from sympy import exp
+    >>> from sympy.abc import x
+
+    Let the equation whose inverse's expansion is to be found be y = f(x)
+    then the first argument to be supplied must be f(x).
+
+    >>> f = exp(x)
+
+    In this case, the inverse series is that of the logarithmic function:
+
+    >>> inversion_series(f,x,0,4)
+    x + (x-1)**3/3 - (x-1)**2/2 - 1
+
+    Functions with vanishing derivative cannot be inverted with this method
+    >>> f = x**2
+    >>> inversion_series(f,x,0,2)
+    ValueError: The inversion formula requires non-zero derivative.
+
+    Note
+    ====
+
+    There is a straightforward derivation using complex analysis and contour integration;
+    the complex formal power series version is a consequence of knowing the formula for polynomials,
+    so the theory of analytic functions may be applied. Actually, the machinery from analytic
+    function theory enters only in a formal way in this proof, in that what is really needed is
+    some property of the formal residue, and a more direct formal proof is available.
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Lagrange_inversion_theorem
+    .. [2] http://mathworld.wolfram.com/LagrangeInversionTheorem.html
+    .. [3] http://www.cfm.brown.edu/people/dobrush/am33/Mathematica/lit.html
+       [4] Abramowitz, M. and Stegun, I.A., 1964. "Handbook of Mathematical Functions", Eq. 3.6.6
+    """
+
+    if f.diff(x,1).subs({x:a}) == 0:
+        raise ValueError("The inversion formula requires non-zero derivative.")
+
+    fa = f.subs({x:a})
+    w = (x-a)/(f-fa)
+    ks = range(1,n)
+    coeffs = [(w**k).diff(x,k-1).limit(x,a) for k in ks ]
+    poly = a+sum(c/factorial(k)*(x-fa)**k for c,k in zip(coeffs,ks))
```

Comment:
It should be possible to compute the inverted series directly from the series for f without any need to use limits and derivatives.

Reply:
A formula exists and is called `reversion`:

https://mathworld.wolfram.com/SeriesReversion.html

but it may not be easy to evaluate it efficiently.

Reply:
There is already a function that does this:
https://github.com/sympy/sympy/blob/ae895046cd7d5fea6194d46cb51776413ae25f1c/sympy/polys/ring_series.py#L575

Reply:
Not quite. That function operates on formal polynomials. Not easy to bridge all use-cases like complicated expressions with many parameters.

As opposed to that, the intention of this PR was to implement Lagrange's Inversion. The PR was also de-facto prapproved before, so basic research was done I suppose?

Reply:
Example:
```python
In [15]: t, z = symbols('t, z')

In [16]: f = exp(z) - 1

In [17]: f.series(z, n=5)
Out[17]: 
     2    3    4        
    z    z    z     ⎛ 5⎞
z + ── + ── + ── + O⎝z ⎠
    2    6    24        

In [18]: solve(f - t, z)
Out[18]: [log(t + 1)]

In [19]: solve(f - t, z)[0].series(t, n=5)
Out[19]: 
     2    3    4        
    t    t    t     ⎛ 5⎞
t - ── + ── - ── + O⎝t ⎠
    2    3    4         

In [20]: from sympy.polys.ring_series import rs_series_reversion

In [21]: R, x, y = ring('x, y', QQ)

In [22]: rs_series_reversion(x + x**2/2 + x**3/6 + x**4/24, x, 5, y)
Out[22]: -1/4*y**4 + 1/3*y**3 - 1/2*y**2 + y
```

Reply:
No doubt that you can invert formal polynomials.
But would that work for complicated expressions which are not polynomial-ring compatible, that was my question.

Because the current expression does.

<img width="1285" alt="image" src="https://user-images.githubusercontent.com/31315784/201396132-3fc78bbe-ddab-460f-9a42-3fcb90d7d11a.png">

Reply:
> The PR was also de-facto prapproved before, so basic research was done I suppose?

No one approved or merged the PR. I personally didn't do any basic research before but I now have (I literally googled "lagrange inversion algorithm" and skimmed the first paper that came up): https://github.com/sympy/sympy/issues/14110#issuecomment-1311876336
The algorithms I found there are not complicated. There is much about efficient large polynomial/matrix multiplication which is irrelevant for SymPy's purposes.

The point though is that those algorithms show how to perform reversion while operating directly on the series and using only simple arithmetic. Algorithm 2 just arranges the series coefficients in a pair of matrices and then multiplies the two matrices.

If it is possible to calculate the series reversion from the forward series without using the `limit` function then that would be a lot better even if a naive algorithm is used for the necessary arithmetic. The `limit` function is one of a class of functions like `integrate`, `solve`, etc that should not be used unless there is no alternative because there are many situations where these functions might necessarily fail (due to the nature of the problems they solve being ill-posed in general). These functions are great for end users but you shouldn't build anything out of them if there is a more direct way to do what is needed.

Reply:
@oscarbenjamin 
Thanks. I know that paper too. Still, this doesn't address my use-case. Look at the example I shared. Expressions in practice will be _polynomials with non-poly expressions_.  Can you handle them in Sympy? How, using a custom domain?

If the concern is only about `limit`, we may want to try adapting the formula a bit.

Reply:
> May I ask what is your motivation for trying to fit it into the algebraic framework?
> Efficiency? Avoiding code redundancy?

There is a bit of both of these things and also the point in my previous post. Note that this is how most things in SymPy work. The heavy algorithms live in the algebraic framework and the high-level functions just manage converting into the algebraic framework.

Reply:
> Look please at my example: expressions in practice will be _polynomials with non-poly expressions_.

If you look at the algorithms described in the linked paper you will see that there is no requirement for the coefficient ring R to be a polynomial ring. It just needs to support `+` and `*` (i.e. all SymPy expressions could be valid).

I'm not sure what limitations there are for the existing `rs_series_reversion` function. I suspect that those limitations are only because it is a low level function that expects to be called with a ring argument. The generators of the ring could still be expressions though e.g.:
```python
In [6]: Poly(cos(t)*x, x)
Out[6]: Poly(cos(t)*x, x, domain='ZZ[cos(t)]')
```

Reply:
> The point though is that those algorithms show how to perform reversion while operating directly on the series and using only simple arithmetic. Algorithm 2 just arranges the series coefficients in a pair of matrices and then multiplies the two matrices.

You missed the part which reverses `f` in the polynomial ring, which will not be well-supported in Sympy for arbitrary coefficients.

Here is the thing: the current proposal does what it promises, for arbitrary expressions.
You have a great idea of connecting it to the algebraic framework, but there are limitations, and we need to check feasibility. The first step should be to replicate the case I shared. Let me give it some more thinking.

Reply:
@oscarbenjamin 
I implemented a variant of Newton updates, which makes n iterations for an n-th order approximation. 
Hence, `limit` is removed, and the performance is improved. 

Furthermore, it works for any expression also beyond the scope of the technical function you found.

I consider this review-ready. 
Would be happy to work on fancy algorithms for inversion too, yet that should be done under a different PR. Here we want to deliver a robust algorithm of broad scope of applicability, not speculate on ring implementations :-)

Reply:
> You missed the part which inverts `f` in the polynomial ring, which will not be well-supported in Sympy with arbitrary coefficients.

I don't immediately have time to review this PR but I just want to pick up on this one point. The algorithm described in the paper presumes that our starting point is a series but in our case the starting point is actually an expression. SymPy already has the capability to compute the series of the reciprocal of an expression:
```python
In [1]: f = exp(x) - 1

In [2]: series(1/f, x)
Out[2]: 
               3      5         
1   1   x     x      x      ⎛ 6⎞
─ - ─ + ── - ─── + ───── + O⎝x ⎠
x   2   12   720   30240        
```
Also you perhaps do not appreciate the extent to which SymPy can make do with the `EX` domain:
https://docs.sympy.org/latest/modules/polys/domainsintro.html#expression-domain
Whatever expressions you imagine would not work well with the algebraic structures used by `ring_series` could at least be represented by the `EX` domain. If the computation does not require testing zero equivalence then the `EX` domain might be slow but is always available.

Reply:
My advice is: let's review this PR, judging by what we said in the beginning: should use minimal operations (add,mul, no limit) and be universal (no ring embeddings). If something is missing, let's make a TODO list. This has been open for a while, and people participating – me too—would like to see light in the tunnel. 

Then I would be happy to work a little bit on more advanced algorithms you suggested. Note that they don't have to bring any gain: may not be faster in practice.